Add new gall guide. (#1249)
urbit/urbit.org: content/docs/userspace/gall-guide/10-scry.md

+++
title = "10. Scries"
weight = 50
template = "doc.html"
+++

In this lesson we'll look at scrying agents, as well as how agents handle such
scries. If you're not at all familiar with performing scries in general, have a
read through the [Scry Guide](/docs/arvo/concepts/scry), as well as the [dotket
rune documentation](/docs/hoon/reference/rune/dot#-dotket).

## Scrying

> Scries are purely read-only requests to vanes and agents which can be initiated

Reviewer: ```suggestion
Scries are read-only requests to vanes and agents which can be initiated
```

from anywhere in any code, during its evaluation. If successful, the requested
data is returned immediately to that point in the code, otherwise it crashes.
This is in contrast to passing around `card`s, which aren't read-only and which
can only be produced when evaluation of the arm is complete, and for which any

> response will come back separately.

Reviewer: ```suggestion
from anywhere in any code, during its evaluation. The requested
data is returned immediately to that point in the code if the read was successful, and crashes otherwise.
This is in contrast to passing around `card`s, which are not read-only,
can only be produced when evaluation of the arm is complete, and responses for which come back asynchronously.
```

Feel like this whole paragraph can be rewritten to be a bit more clear/succinct on the difference here. Arguably scries shouldn't be compared to cards at all, because it's an apples vs oranges situation.

Technically, Arvo only lets you scry vanes. When you scry a Gall agent, you

> actually scry Gall itself, which interprets the request, runs it on the

Reviewer: You can actually scry into arvo itself too! This may be more accurately phrased as "Most of the time, scry requests are handled by Arvo, which routes the request to the appropriate vane."

specified agent, and then returns the result. Scries are performed with the
[dotket](/docs/hoon/reference/rune/dot#-dotket) (`.^`) rune. Here's a summary of
their format:

![scry summary diagram](https://media.urbit.org/docs/arvo/scry-diagram-v2.svg)

> A note on `care`s: Cares are most fully exploited by Clay, where they specify

Reviewer: ```suggestion
A note on `care`s: Cares are most carefully implemented by Clay, where they specify
```

Alternatively, "most fully supported".

`%d`, `%e`, `%u` and `%x`.

#### `%d`

A scry to Gall with a `%d` `care` and no `path` will produce the `desk` in which
the specified agent resides. For example:

```
> .^(desk %gd /=hark-store=)
%garden
> .^(desk %gd /=hood=)
%base
```

#### `%e`

A scry to Gall with a `%e` `care`, a `desk` rather than agent in the `desk`
field of the above diagram, and no path, will produce a set of all installed
agents on that desk and their status. For example:

```
> .^((set [=dude:gall live=?]) %ge /=garden=)
{ [dude=%hark-system-hook live=%.y]
  [dude=%treaty live=%.y]
  [dude=%docket live=%.y]
  [dude=%settings-store live=%.y]
  [dude=%hark-store live=%.y]
}
```

#### `%u`

A scry to Gall with a `%u` `care` and no `path` will check whether or not the
specified agent is installed and running:

```
> .^(? %gu /=btc-wallet=)
%.y
> .^(? %gu /=btc-provider=)
%.n
> .^(? %gu /=foobar=)
%.n
```

#### `%x`

A scry to Gall with a `%x` `care` will be passed to the agent for handling. Gall
handles `%x` specially, and expects an extra field at the end of the `path` that
specifies the `mark` to return. Gall will take the data produced by the

> specified endpoint and try convert it to the given mark, crashing if the mark

Reviewer 2: ```suggestion
specified endpoint and try to convert it to the given mark, crashing if the mark
```

```
> =store -build-file /=landscape=/sur/graph-store/hoon

> .^(update:store %gx /=graph-store=/keys/noun)
[p=~2021.11.18..10.50.41..c914 q=[%keys resources={[entity=~zod name=%dm-inbox]}]]

> (crip (en-json:html .^(json %gx /=graph-store=/keys/json)))
'{"graph-update":{"keys":[{"name":"dm-inbox","ship":"zod"}]}}'
```

The majority of Gall agents simply take `%x` `care`s in their scry endpoints,
but in principle it's possible for a Gall agent to define a scry endpoint that
takes any one of the `care`s listed in the diagram above. An agent's scry
endpoints are defined in its `on-peek` arm, which we'll look at next.

## Handling scries

When a scry is performed on a Gall agent, Gall will strip out some extraneous
parts, and deliver it to the agent's `on-peek` arm as a `path`. The `path` will
only have two components from the diagram above: The _care_ and the _path_. For
example, a scry of `.^(update:store %gx /=graph-store=/keys/noun)` will come
into the `on-peek` arm of `%graph-store` as `/x/keys`.

The `on-peek` arm produces a `(unit (unit cage))`. The reason for the double
`unit` is that Arvo interprets `~` to mean the scry endpoint doesn't exist at

> all, and interprets `[~ ~]` to mean the endpoint exists but the requested data

Reviewer: I always thought this was more or less the other way around. ie `~` for "no answer", and `[~ ~]` for "invalid/doesn't exist"...

does not. In either case the dotket expression which initiated the scry will
crash. The `cage` will contain the actual data to return.

> The `on-peek` arm, therefore, begins like so:

Reviewer 2: ```suggestion
An ordinary `on-peek` arm, therefore, begins like so:
```

```hoon
++  on-peek
  |=  =path
  ^-  (unit (unit cage))
  ....
```

Typically, you'd handle the `path` similarly to `on-watch`, as we discussed in
the lesson on subscriptions. You'd use something like a wutlus expression to

> test the type of the `path`, defining your scry endpoints like so:

Reviewer 2: ```suggestion
test the value of the `path`, defining your scry endpoints like so:
```

```hoon
?+    path  (on-peek:def path)
    [%x %some %path ~]  ....
    [%x %foo ~]    ....
    [%x %blah @ ~]
  =/  =ship  (slav %p i.t.t.path)
  .....
....
```

Each endpoint would then compose the `(unit (unit cage))`. The simplest way to
format it is like:

```hoon
``noun+!>('some data')
```

If it requires a more complex expression to retrieve or compose the data, you
can do something like:

```hoon
:-  ~  :-  ~  :-  %some-mark

> !>  ^-  some-type

Reviewer: What's wrong with `:^`?

:+  'foo'
  'bar'
'baz'
```

Previously we discussed custom `mark` files. Such mark files are most commonly
used when the data might be accessed through Eyre's HTTP API, and therefore
required JSON conversion methods. We'll cover such things in part two of the
Gall Guide, but note that if that's the case for your agent, you may wish to
also have your scry endpoints return data with your custom `mark` so it can
easily be converted to JSON when accessed from the web.

In some cases, typically with scry `path`s that contain wildcards like the `[%x %blah @ ~]` example above, your agent may not always be able to find the
requested data. In such cases, you can just produce a cell of `[~ ~]` for the
`(unit (unit cage))`. Keep in mind, however, that this will result in a crash
for the dotket expression which initiated the scry. In some cases you may want
that, but in other cases you may not, so instead you could wrap the data inside
the `vase` in a `unit` and have _that_ be null instead. It all depends on the

> needs of your particular application.

Reviewer: ```suggestion
needs of your particular application and its clients.
```

`@p` if it exists in the map, or else return `[~ ~]` to indicate the data
doesn't exist, producing a crash in the dotket expression.

Let's try it out. Save the agent above as `/app/peeker.hoon` in the `%base`
desk, `|commit %base` and start the agent with `|rein %base [& %peeker]`.

First, let's add some data to the map:

```
> :peeker [~zod 'foo']
>=
> :peeker [~nut 'bar']
>=
> :peeker [~wet 'baz']
>=
```

Now if we use `dbug` to inspect the state, we'll see the data has been added:

```
>   {[p=~wet q='baz'] [p=~nut q='bar'] [p=~zod q='foo']}
> :peeker +dbug [%state %data]
>=
```

Next, let's try the `/x/all` scry endpoint:

```
> .^((map @p @t) %gx /=peeker=/all/noun)
{[p=~wet q='baz'] [p=~nut q='bar'] [p=~zod q='foo']}
```

The `/x/has/[ship]` endpoint:

```
> .^(? %gx /=peeker=/has/~zod/noun)
%.y
> .^(? %gx /=peeker=/has/~wet/noun)
%.y
> .^(? %gx /=peeker=/has/~nes/noun)
%.n
```

And finally, the `/x/get/[ship]` endpoint:

```
> .^(@t %gx /=peeker=/get/~zod/noun)
'foo'
> .^(@t %gx /=peeker=/get/~wet/noun)
'baz'
```

We'll now try scrying for a ship that doesn't exist in the map. Note that the
resulting crash will wipe the dojo's subject, so don't try this if you've got

> anything pinned there that you want to keep.

Reviewer: Might be worth noting that this is not intended/necessary behavior lol.

```
~zod:dojo> .^(@t %gx /=peeker=/get/~nes/noun)
crash!
```

## Summary

- Scries are read-only requests to vanes or agents which can be done inside any
  code, during its evaluation.
- Scries are performed with the dotket (`.^`) rune.
- Scries will fail if the scry endpoint does not exist, the requested data does
  not exist, or the data does not nest in the return type specified.
- Scries can only be performed on the local ship, not on remote ships.
- Gall scries with an agent name in the `desk` field will be passed to that
  agent's `on-peek` arm for handling.

> - Gall scries with a `%x` `care` take a `mark` at the end of the scry path,

Reviewer 2: ```suggestion
- Gall scries with a `%x` `care` take a `mark` at the end of the scry `path`,
```

>   telling gall to convert the data returned by the scry endpoint to the mark

Reviewer 2: ```suggestion
  telling Gall to convert the data returned by the scry endpoint to the mark
```

  specified.

> - The `on-peek` arm takes a `path` with the `care` in the head and the path part

Reviewer 2: ```suggestion
- The `on-peek` arm takes a `path` with the `care` in the head and the `path` part
```
